pkg/prometheus/agent/operator.go

```diff
@@ -0,0 +1,1390 @@
+		}
+	})
+	if err != nil {
+		level.Error(c.logger).Log(
+			"msg", "listing all Prometheus instances from cache failed",
+			"err", err,
+		)
+	}
+
+}
+
+func (c *Operator) handleMonitorNamespaceUpdate(oldo, curo interface{}) {
+	old := oldo.(*v1.Namespace)
+	cur := curo.(*v1.Namespace)
+
+	level.Debug(c.logger).Log("msg", "update handler", "namespace", cur.GetName(), "old", old.ResourceVersion, "cur", cur.ResourceVersion)
+
+	// Periodic resync may resend the Namespace without changes
+	// in-between.
+	if old.ResourceVersion == cur.ResourceVersion {
+		return
+	}
+
+	level.Debug(c.logger).Log("msg", "Monitor namespace updated", "namespace", cur.GetName())
+	c.metrics.TriggerByCounter("Namespace", operator.UpdateEvent).Inc()
+
+	// Check for Prometheus Agent instances selecting ServiceMonitors, PodMonitors,
+	// and Probes in the namespace.
+	err := c.promInfs.ListAll(labels.Everything(), func(obj interface{}) {
+		p := obj.(*monitoringv1alpha1.PrometheusAgent)
+
+		for name, selector := range map[string]*metav1.LabelSelector{
+			"PodMonitors":     p.Spec.PodMonitorNamespaceSelector,
+			"Probes":          p.Spec.ProbeNamespaceSelector,
+			"ServiceMonitors": p.Spec.ServiceMonitorNamespaceSelector,
+		} {
+
+			sync, err := k8sutil.LabelSelectionHasChanged(old.Labels, cur.Labels, selector)
+			if err != nil {
+				level.Error(c.logger).Log(
+					"err", err,
+					"name", p.Name,
+					"namespace", p.Namespace,
+					"subresource", name,
+				)
+				return
+			}
+
+			if sync {
+				c.rr.EnqueueForReconciliation(p)
+				return
+			}
+		}
+	})
+	if err != nil {
+		level.Error(c.logger).Log(
+			"msg", "listing all Prometheus Agent instances from cache failed",
+			"err", err,
+		)
+	}
+}
+
+// checkAgentAuthorization checks if the operator has access to PrometheusAgent CRD
+// It checks individual namespaces if an allowlist was provided, otherwise it checks cluster-wide.
+func (c *Operator) checkAgentAuthorization(ctx context.Context) (bool, error) {
+	verbs := []string{"get", "list", "watch", "update"}
+	var ssar *authv1.SelfSubjectAccessReview
+	var ssarResponse *authv1.SelfSubjectAccessReview
+	var err error
+
+	// If PrometheusAllowList is empty, 'ns' will be an empty string,
+	// making the check cluster-wide.
+	for ns := range c.config.Namespaces.PrometheusAllowList {
+		for _, verb := range verbs {
+			ssar = &authv1.SelfSubjectAccessReview{
+				Spec: authv1.SelfSubjectAccessReviewSpec{
+					ResourceAttributes: &authv1.ResourceAttributes{
+						Verb:      verb,
+						Group:     "monitoring.coreos.com/v1alpha1",
+						Resource:  "prometheusagents",
+						Namespace: ns,
+					},
+				},
+			}
+			ssarResponse, err = c.kclient.AuthorizationV1().SelfSubjectAccessReviews().Create(ctx, ssar, metav1.CreateOptions{})
+			if err != nil {
+				return false, err
+			}
+			if !ssarResponse.Status.Allowed {
+				return false, nil
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func (c *Operator) checkCRDInstallation(ctx context.Context) error {
```

Comment:
After more thinking, I don't feel that checking the CRD presence is the best approach because it requires additional permissions for the prometheus-operator account (e.g. get permissions on `/apis/apiextensions.k8s.io/v1/customresourcedefinitions`) which is one more thing to setup for the operator admin before they can use the PrometheusAgent CRD.
Maybe checking that the operator's service account has all the required permissions is good enough? If you've updated the SA permissions, you've also probably deployed the PrometheusAgent CRD.

Reply:
Yeah, I started thinking the same when implementing the e2e tests. Good idea with the serviceaccount

Reply:
That means that the `SelfSubjectAccessReview` is enough, right?